protoc-gen-openapi/openapiGenerator.go

```diff
@@ -109,95 +113,154 @@ func (g *openapiGenerator) generatePerPackageOutput(filesToGen map[*protomodel.F
 
-		for _, field := range message.Fields {
-			s.WithProperty(field.GetName(), g.fieldType(field))
+	for _, message := range messages {
+		if message.Parent == nil {
+			g.generateMessage(message, allSchemas)
 		}
 	}
 
-	var schemas allSchemas
-	for _, schema := range g.schemas {
-		schemas.Schemas = append(schemas.Schemas, schema)
+	for _, enum := range enums {
+		// when there is no parent to the enum.
+		if len(enum.QualifiedName()) == 1 {
+			g.generateEnum(enum, allSchemas)
+		}
 	}
 
+	c := openapi3.NewComponents()
+	c.Schemas = allSchemas
+
 	g.buffer.Reset()
-	b, _ := yaml.Marshal(&schemas)
+	b, _ := json.MarshalIndent(c, "", "  ")
 	g.buffer.Write(b)
 
 	return plugin.CodeGeneratorResponse_File{
-		Name:    proto.String(name + ".yaml"),
+		Name:    proto.String(name + ".json"),
 		Content: proto.String(g.buffer.String()),
 	}
 }
 
+func (g *openapiGenerator) generateMessage(message *protomodel.MessageDescriptor, allSchemas map[string]*openapi3.SchemaRef) {
+	o := g.generateMessageSchema(message)
+	if o != nil {
+		allSchemas[g.absoluteName(message)] = o.NewRef()
+	}
+}
+
+func (g *openapiGenerator) generateMessageSchema(message *protomodel.MessageDescriptor) *openapi3.Schema {
+	// skip MapEntry message because we handle map using the map's repeated field.
+	if message.GetOptions().GetMapEntry() {
+		return nil
+	}
+	o := openapi3.NewObjectSchema()
+	o.Description = g.generateDescription(message)
+	oneof := make(map[int32][]*openapi3.Schema)
+	for _, field := range message.Fields {
+		// skip deprecated field as it is not supported by Kubernetes yet.
+		if field.GetOptions().GetDeprecated() {
+			continue
+		}
+		if field.OneofIndex == nil {
+			o.WithProperty(field.GetName(), g.fieldType(field))
+		} else {
+			oneof[*field.OneofIndex] = append(oneof[*field.OneofIndex], g.fieldType(field))
+		}
+	}
+	for k, v := range oneof {
+		o.WithProperty(message.GetOneofDecl()[k].GetName(), openapi3.NewOneOfSchema(v...))
+	}
+	return o
+}
+
+func (g *openapiGenerator) generateEnum(enum *protomodel.EnumDescriptor, allSchemas map[string]*openapi3.SchemaRef) {
+	o := g.generateEnumSchema(enum)
+	if o != nil {
+		allSchemas[g.absoluteName(enum)] = o.NewRef()
+	}
+}
+
+func (g *openapiGenerator) generateEnumSchema(enum *protomodel.EnumDescriptor) *openapi3.Schema {
+	o := openapi3.NewStringSchema()
+	o.Description = g.generateDescription(enum)
+	values := enum.GetValue()
+	enumNames := make([]string, len(values))
+	for i, v := range values {
+		enumNames[i] = v.GetName()
+	}
+	o.WithEnum(enumNames)
+	return o
+}
+
+func (g *openapiGenerator) absoluteName(desc protomodel.CoreDesc) string {
+	typeName := protomodel.DottedName(desc)
+	return desc.PackageDesc().Name + "." + typeName
+}
+
+// converts the leading comment or the description of the proto to a single line of description.
+func (g *openapiGenerator) generateDescription(desc protomodel.CoreDesc) string {
+	d := desc.Location().GetLeadingComments()
+	return strings.Join(strings.Fields(d), " ")
```

Comment:
Won't this mangle a lot of comments? Protos comments are assumed to be markdown, and we use a lot of code blocks that won't fare well if turned into a single line. Is there a better option here?

Reply:
as discussed offline, we will just use the first paragraph of the leading comments.